File: mcp/src/blender/tools/advanced/physics.js
// Advanced physics tools — cloth, rigid body, soft body, fluid, collision, baking

import { sendToBlender } from "../../blender-bridge.js";

export const physicsTools = [
  {
    name: "blender_physics_cloth_add",
    description:
      "Add a Cloth physics modifier to a mesh object, enabling fabric simulation. " +
      "Optionally apply a material preset (COTTON, SILK, LEATHER, DENIM, RUBBER) that configures " +
      "realistic stiffness, mass, and quality settings for that material type. " +
      "Example call: { \"object_name\": \"Plane\", \"preset\": \"SILK\" }. " +
      "The object must be a mesh. After adding, the cloth will simulate on playback.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the mesh object to add cloth physics to.",
        },
        preset: {
          type: "string",
          enum: ["COTTON", "SILK", "LEATHER", "DENIM", "RUBBER"],
          description:
            "Optional cloth material preset. Must be UPPERCASE. " +
            "COTTON: medium stiffness, light. " +
            "SILK: very soft, light, drapes fluidly. " +
            "LEATHER: stiff, heavy. " +
            "DENIM: medium-stiff, medium weight. " +
            "RUBBER: elastic, heavy.",
        },
      },
      required: ["object_name"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/cloth-add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_cloth_configure",
    description:
      "Configure parameters of an existing Cloth modifier on a mesh object. " +
      "Pass a params object with any combination of: mass (kg), tension_stiffness, " +
      "compression_stiffness, bending_stiffness, quality (simulation steps per frame), " +
      "air_damping, velocity_max. Only the keys provided will be updated. " +
      "Example call: { \"object_name\": \"Plane\", \"params\": { \"mass\": 0.3, \"tension_stiffness\": 15 } }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the mesh object with a Cloth modifier.",
        },
        params: {
          type: "object",
          description:
            "IMPORTANT: Must be a JSON object, NOT a string. " +
            "Cloth settings to apply. Supported keys: " +
            "mass (float, kg), " +
            "tension_stiffness (float), " +
            "compression_stiffness (float), " +
            "bending_stiffness (float), " +
            "quality (int, simulation steps/frame), " +
            "air_damping (float), " +
            "velocity_max (float).",
          properties: {
            mass: { type: "number", description: "Cloth mass in kg." },
            tension_stiffness: {
              type: "number",
              description: "Resistance to stretching.",
            },
            compression_stiffness: {
              type: "number",
              description: "Resistance to compression.",
            },
            bending_stiffness: {
              type: "number",
              description: "Resistance to bending.",
            },
            quality: {
              type: "integer",
              description: "Simulation steps per frame (higher = more accurate).",
            },
            air_damping: {
              type: "number",
              description: "Air resistance / drag coefficient.",
            },
            velocity_max: {
              type: "number",
              description: "Maximum velocity clamping.",
            },
          },
        },
      },
      required: ["object_name", "params"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/cloth-configure", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_cloth_pin",
    description:
      "Assign a vertex group as the cloth pin group, fixing those vertices in space " +
      "so they do not move during simulation. The vertex group must already exist on the object. " +
      "Useful for hanging fabric from a fixed attachment point.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the mesh object with a Cloth modifier.",
        },
        vertex_group: {
          type: "string",
          description:
            "Name of the vertex group to use as the pin group. " +
            "Vertices in this group will be held in place during simulation.",
        },
      },
      required: ["object_name", "vertex_group"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/cloth-pin", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_rigid_body_add",
    description:
      "Add a Rigid Body physics simulation to an object. " +
      "ACTIVE objects are fully simulated (they move, fall, collide). " +
      "PASSIVE objects are immovable collision obstacles (e.g. floor, walls). " +
      "Optionally set the collision shape and mass.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object to add rigid body physics to.",
        },
        type: {
          type: "string",
          enum: ["ACTIVE", "PASSIVE"],
          description:
            "Must be UPPERCASE. ACTIVE: object is simulated and moves freely. " +
            "PASSIVE: object is a static collision obstacle.",
        },
        shape: {
          type: "string",
          enum: [
            "BOX",
            "SPHERE",
            "CAPSULE",
            "CYLINDER",
            "CONE",
            "CONVEX_HULL",
            "MESH",
          ],
          description:
            "Collision shape. Must be UPPERCASE. BOX/SPHERE/CAPSULE/CYLINDER/CONE are fast primitives. " +
            "CONVEX_HULL approximates the mesh shape. MESH is exact but slow.",
        },
        mass: {
          type: "number",
          description: "Mass in kg for ACTIVE objects. Defaults to 1.0.",
        },
      },
      required: ["object_name", "type"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/rigid-body-add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_rigid_body_configure",
    description:
      "Configure parameters of an existing Rigid Body on an object. " +
      "Pass a params object with any combination of: mass, friction, restitution " +
      "(bounciness), linear_damping, angular_damping, collision_shape, kinematic " +
      "(animated), use_deactivation. Only provided keys are updated. " +
      "Example call: { \"object_name\": \"Cube\", \"params\": { \"mass\": 2.0, \"friction\": 0.5, \"kinematic\": false } }.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object with a rigid body.",
        },
        params: {
          type: "object",
          description:
            "IMPORTANT: Must be a JSON object, NOT a string. " +
            "Rigid body parameters to apply. Supported keys: " +
            "mass (float, kg), " +
            "friction (float, 0-1), " +
            "restitution (float, 0-1, bounciness), " +
            "linear_damping (float, 0-1), " +
            "angular_damping (float, 0-1), " +
            "collision_shape (string: BOX/SPHERE/CAPSULE/CYLINDER/CONE/CONVEX_HULL/MESH), " +
            "kinematic (bool, treat as animated), " +
            "use_deactivation (bool).",
          properties: {
            mass: { type: "number", description: "Mass in kg." },
            friction: { type: "number", description: "Surface friction (0-1)." },
            restitution: {
              type: "number",
              description: "Bounciness / restitution (0-1).",
            },
            linear_damping: {
              type: "number",
              description: "Linear velocity damping (0-1).",
            },
            angular_damping: {
              type: "number",
              description: "Rotational velocity damping (0-1).",
            },
            collision_shape: {
              type: "string",
              description: "Collision shape type.",
            },
            kinematic: {
              type: "boolean",
              description:
                "If true, object is animated (not fully simulated). " +
                "IMPORTANT: Must be a JSON boolean (true or false), NOT a string.",
            },
            use_deactivation: {
              type: "boolean",
              description:
                "Allow the object to deactivate when at rest. " +
                "IMPORTANT: Must be a JSON boolean (true or false), NOT a string.",
            },
          },
        },
      },
      required: ["object_name", "params"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/rigid-body-configure", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_soft_body_add",
    description:
      "Add a Soft Body physics modifier to a mesh object, enabling elastic/jelly-like " +
      "deformation simulation. Optionally configure initial parameters such as mass, " +
      "goal_stiffness, goal_damping, and pull/push stiffness.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the mesh object to add soft body physics to.",
        },
        params: {
          type: "object",
          description:
            "IMPORTANT: Must be a JSON object, NOT a string. " +
            "Optional soft body settings. Supported keys: " +
            "mass (float, per-vertex mass), " +
            "goal_stiffness (float, 0-1, how strongly verts return to rest shape), " +
            "goal_damping (float, 0-1), " +
            "pull (float, edge pull stiffness), " +
            "push (float, edge push stiffness), " +
            "damping (float, edge damping), " +
            "bend (float, bending stiffness).",
          properties: {
            mass: { type: "number", description: "Per-vertex mass." },
            goal_stiffness: {
              type: "number",
              description: "Goal stiffness (0-1).",
            },
            goal_damping: {
              type: "number",
              description: "Goal damping (0-1).",
            },
            pull: { type: "number", description: "Edge pull stiffness." },
            push: { type: "number", description: "Edge push stiffness." },
            damping: { type: "number", description: "Edge damping." },
            bend: { type: "number", description: "Bending stiffness." },
          },
        },
      },
      required: ["object_name"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/soft-body-add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_collision_add",
    description:
      "Add a Collision modifier to a mesh object, making it act as a collision " +
      "obstacle for cloth and soft body simulations. Optionally set the outer thickness " +
      "and friction coefficient.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the mesh object to add collision physics to.",
        },
        thickness: {
          type: "number",
          description:
            "Outer collision thickness in Blender units. " +
            "Controls how far from the surface cloth/soft bodies are repelled. Defaults to 0.025.",
        },
        friction: {
          type: "number",
          description:
            "Surface friction coefficient (0 = frictionless, 1 = very sticky). Defaults to 5.0.",
        },
      },
      required: ["object_name"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/collision-add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_fluid_add",
    description:
      "Add a Fluid (Mantaflow) physics modifier to an object and set its role. " +
      "DOMAIN defines the simulation bounding box (one per scene). " +
      "FLOW is a fluid emitter (source or sink). " +
      "EFFECTOR creates obstacles or force fields within the simulation.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description: "Name of the object to add fluid physics to.",
        },
        type: {
          type: "string",
          enum: ["DOMAIN", "FLOW", "EFFECTOR"],
          description:
            "Fluid role. Must be UPPERCASE. " +
            "DOMAIN — defines the simulation volume (bounding box). Only one domain per scene. " +
            "FLOW — fluid emitter or drain. " +
            "EFFECTOR — obstacle or guide within the simulation.",
        },
      },
      required: ["object_name", "type"],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/fluid-add", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_bake",
    description:
      "Bake physics simulations to the cache so they play back deterministically. " +
      "If object_name is provided, the scene is set up with that object active before baking. " +
      "If omitted, all physics simulations in the scene are baked. " +
      "Optionally override the scene's start and end frame for the bake range.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description:
            "Optional: name of the object to focus baking on. " +
            "If omitted, all simulations in the scene are baked.",
        },
        start_frame: {
          type: "integer",
          description:
            "Optional: first frame of the bake range. Overrides the scene's current frame_start.",
        },
        end_frame: {
          type: "integer",
          description:
            "Optional: last frame of the bake range. Overrides the scene's current frame_end.",
        },
      },
      required: [],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/bake", args);
      return JSON.stringify(result, null, 2);
    },
  },

  {
    name: "blender_physics_free_bake",
    description:
      "Free (clear) cached physics bake data so simulations will be re-evaluated " +
      "on the next playback or bake. If object_name is provided, that object is made active " +
      "before freeing. If omitted, all bake caches in the scene are freed.",
    inputSchema: {
      type: "object",
      properties: {
        object_name: {
          type: "string",
          description:
            "Optional: name of the object whose bake cache to free. " +
            "If omitted, all simulation caches in the scene are freed.",
        },
      },
      required: [],
    },
    handler: async (args) => {
      const result = await sendToBlender("physics/free-bake", args);
      return JSON.stringify(result, null, 2);
    },
  },
];

File: src/scripts/mixar/modules/mcp_bridge/blender/handlers/advanced/physics.py
"""
Advanced physics handlers for Blender MCP Bridge.
Provides: physics/cloth-add, physics/cloth-configure, physics/cloth-pin,
          physics/rigid-body-add, physics/rigid-body-configure,
          physics/soft-body-add, physics/collision-add, physics/fluid-add,
          physics/bake, physics/free-bake
"""

import bpy
from ...utils.response import ok_response, error_response, not_found, coerce_value
from ...utils.context_helpers import ensure_context_for_object, temp_override
from .. import register_handler


# ─── Constants ──────────────────────────────────────────────────────────────────

CLOTH_PRESETS = {
    "COTTON": {
        "quality": 5,
        "mass": 0.3,
        "tension_stiffness": 15,
        "compression_stiffness": 15,
        "bending_stiffness": 0.5,
    },
    "SILK": {
        "quality": 5,
        "mass": 0.15,
        "tension_stiffness": 5,
        "compression_stiffness": 5,
        "bending_stiffness": 0.05,
    },
    "LEATHER": {
        "quality": 15,
        "mass": 0.4,
        "tension_stiffness": 80,
        "compression_stiffness": 80,
        "bending_stiffness": 150,
    },
    "DENIM": {
        "quality": 12,
        "mass": 0.3,
        "tension_stiffness": 40,
        "compression_stiffness": 40,
        "bending_stiffness": 10,
    },
    "RUBBER": {
        "quality": 7,
        "mass": 3.0,
        "tension_stiffness": 15,
        "compression_stiffness": 15,
        "bending_stiffness": 25,
    },
}

# Cloth settings keys that map directly to ClothSettings attributes
CLOTH_CONFIGURABLE_KEYS = {
    "mass",
    "tension_stiffness",
    "compression_stiffness",
    "bending_stiffness",
    "quality",
    "air_damping",
    "velocity_max",
}

# Soft body settings keys
SOFTBODY_CONFIGURABLE_KEYS = {
    "mass",
    "goal_stiffness",
    "goal_damping",
    "pull",
    "push",
    "damping",
    "bend",
}


# ─── Helpers ────────────────────────────────────────────────────────────────────

def _get_object(name):
    """Return (obj, None) or (None, error_response) for any object."""
    obj = bpy.data.objects.get(name)
    if obj is None:
        return None, not_found(name)
    return obj, None


def _get_mesh_object(name):
    """Return (obj, None) or (None, error_response) — enforces MESH type."""
    obj = bpy.data.objects.get(name)
    if obj is None:
        return None, not_found(name)
    if obj.type != "MESH":
        return None, error_response(
            f"Object '{name}' is of type '{obj.type}', expected 'MESH'."
        )
    return obj, None


# ─── Handlers ───────────────────────────────────────────────────────────────────

def _handle_cloth_add(params):
    """
    Add a Cloth modifier to a mesh object, optionally applying a material preset.
    Route: POST /api/physics/cloth-add
    """
    object_name = params.get("object_name")
    preset = params.get("preset")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    obj, err = _get_mesh_object(object_name)
    if err:
        return err

    if preset and preset.upper() not in CLOTH_PRESETS:
        return error_response(
            f"Unknown preset '{preset}'. Valid: {', '.join(CLOTH_PRESETS.keys())}."
        )

    try:
        ensure_context_for_object(obj)

        # Check if Cloth modifier already exists
        existing = obj.modifiers.get("Cloth")
        if existing:
            return error_response(
                f"Object '{object_name}' already has a Cloth modifier. "
                "Use physics/cloth-configure to update settings."
            )

        with temp_override("VIEW_3D"):
            bpy.ops.object.modifier_add(type="CLOTH")

        cloth_mod = obj.modifiers.get("Cloth")
        if cloth_mod is None:
            return error_response("Failed to add Cloth modifier.")

        settings = cloth_mod.settings
        applied_preset = None

        if preset:
            preset_upper = preset.upper()
            p = CLOTH_PRESETS[preset_upper]
            settings.quality = p["quality"]
            settings.mass = p["mass"]
            settings.tension_stiffness = p["tension_stiffness"]
            settings.compression_stiffness = p["compression_stiffness"]
            settings.bending_stiffness = p["bending_stiffness"]
            applied_preset = preset_upper

        return ok_response({
            "object_name": obj.name,
            "modifier": "Cloth",
            "preset": applied_preset,
            "settings": {
                "quality": settings.quality,
                "mass": settings.mass,
                "tension_stiffness": settings.tension_stiffness,
                "compression_stiffness": settings.compression_stiffness,
                "bending_stiffness": settings.bending_stiffness,
            },
        })

    except Exception as e:
        return error_response(f"Failed to add Cloth modifier to '{object_name}': {e}")


def _handle_cloth_configure(params):
    """
    Configure an existing Cloth modifier's settings on a mesh object.
    Route: POST /api/physics/cloth-configure
    """
    object_name = params.get("object_name")
    cloth_params = params.get("params") or {}

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not cloth_params:
        return error_response("Parameter 'params' must be a non-empty object.")

    obj, err = _get_mesh_object(object_name)
    if err:
        return err

    cloth_mod = obj.modifiers.get("Cloth")
    if cloth_mod is None:
        return error_response(
            f"Object '{object_name}' has no Cloth modifier. "
            "Use physics/cloth-add first."
        )

    try:
        settings = cloth_mod.settings
        applied = {}
        skipped = {}

        for key, value in cloth_params.items():
            if key in CLOTH_CONFIGURABLE_KEYS:
                if hasattr(settings, key):
                    current = getattr(settings, key)
                    setattr(settings, key, coerce_value(current, value))
                    applied[key] = value
                else:
                    skipped[key] = f"attribute '{key}' not found on ClothSettings"
            else:
                skipped[key] = "unsupported key"

        return ok_response({
            "object_name": obj.name,
            "applied": applied,
            "skipped": skipped,
        })

    except Exception as e:
        return error_response(f"Failed to configure Cloth on '{object_name}': {e}")


def _handle_cloth_pin(params):
    """
    Assign a vertex group as the cloth pin group.
    Route: POST /api/physics/cloth-pin
    """
    object_name = params.get("object_name")
    vertex_group = params.get("vertex_group")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not vertex_group:
        return error_response("Parameter 'vertex_group' is required.")

    obj, err = _get_mesh_object(object_name)
    if err:
        return err

    cloth_mod = obj.modifiers.get("Cloth")
    if cloth_mod is None:
        return error_response(
            f"Object '{object_name}' has no Cloth modifier. "
            "Use physics/cloth-add first."
        )

    if vertex_group not in obj.vertex_groups:
        return error_response(
            f"Vertex group '{vertex_group}' does not exist on '{object_name}'. "
            f"Available groups: {[g.name for g in obj.vertex_groups]}"
        )

    try:
        cloth_mod.settings.vertex_group_mass = vertex_group

        return ok_response({
            "object_name": obj.name,
            "pin_vertex_group": vertex_group,
        })

    except Exception as e:
        return error_response(f"Failed to set cloth pin group on '{object_name}': {e}")


def _handle_rigid_body_add(params):
    """
    Add a Rigid Body to an object.
    Route: POST /api/physics/rigid-body-add
    """
    object_name = params.get("object_name")
    rb_type = params.get("type", "ACTIVE").upper()
    shape = params.get("shape")
    mass = params.get("mass")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    valid_types = ("ACTIVE", "PASSIVE")
    if rb_type not in valid_types:
        return error_response(
            f"Unknown rigid body type '{rb_type}'. Valid: {', '.join(valid_types)}."
        )

    valid_shapes = ("BOX", "SPHERE", "CAPSULE", "CYLINDER", "CONE", "CONVEX_HULL", "MESH")
    if shape and shape.upper() not in valid_shapes:
        return error_response(
            f"Unknown collision shape '{shape}'. Valid: {', '.join(valid_shapes)}."
        )

    obj, err = _get_object(object_name)
    if err:
        return err

    if obj.rigid_body is not None:
        return error_response(
            f"Object '{object_name}' already has a rigid body. "
            "Use physics/rigid-body-configure to update settings."
        )

    try:
        ensure_context_for_object(obj)

        with temp_override("VIEW_3D"):
            bpy.ops.rigidbody.object_add(type=rb_type)

        rb = obj.rigid_body
        if rb is None:
            return error_response(
                f"Failed to add rigid body to '{object_name}'. "
                "Ensure the scene has a Rigid Body World (Object > Rigid Body > Add)."
            )

        if shape:
            rb.collision_shape = shape.upper()

        if mass is not None:
            rb.mass = float(mass)

        return ok_response({
            "object_name": obj.name,
            "type": rb.type,
            "collision_shape": rb.collision_shape,
            "mass": rb.mass,
        })

    except Exception as e:
        return error_response(f"Failed to add rigid body to '{object_name}': {e}")


def _handle_rigid_body_configure(params):
    """
    Configure an existing Rigid Body's settings on an object.
    Route: POST /api/physics/rigid-body-configure
    """
    object_name = params.get("object_name")
    rb_params = params.get("params") or {}

    if not object_name:
        return error_response("Parameter 'object_name' is required.")
    if not rb_params:
        return error_response("Parameter 'params' must be a non-empty object.")

    obj, err = _get_object(object_name)
    if err:
        return err

    rb = obj.rigid_body
    if rb is None:
        return error_response(
            f"Object '{object_name}' has no rigid body. "
            "Use physics/rigid-body-add first."
        )

    try:
        applied = {}
        skipped = {}

        for key, value in rb_params.items():
            if hasattr(rb, key):
                try:
                    current = getattr(rb, key)
                    setattr(rb, key, coerce_value(current, value))
                    applied[key] = value
                except Exception as set_err:
                    skipped[key] = str(set_err)
            else:
                skipped[key] = f"attribute '{key}' not found on RigidBodyObject"

        return ok_response({
            "object_name": obj.name,
            "applied": applied,
            "skipped": skipped,
        })

    except Exception as e:
        return error_response(f"Failed to configure rigid body on '{object_name}': {e}")


def _handle_soft_body_add(params):
    """
    Add a Soft Body modifier to a mesh object.
    Route: POST /api/physics/soft-body-add
    """
    object_name = params.get("object_name")
    sb_params = params.get("params") or {}

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    obj, err = _get_mesh_object(object_name)
    if err:
        return err

    if obj.modifiers.get("Softbody") is not None:
        return error_response(
            f"Object '{object_name}' already has a Soft Body modifier."
        )

    try:
        ensure_context_for_object(obj)

        with temp_override("VIEW_3D"):
            bpy.ops.object.modifier_add(type="SOFT_BODY")

        sb_mod = obj.modifiers.get("Softbody")
        if sb_mod is None:
            return error_response("Failed to add Soft Body modifier.")

        settings = sb_mod.settings
        applied = {}

        for key, value in sb_params.items():
            if key in SOFTBODY_CONFIGURABLE_KEYS and hasattr(settings, key):
                current = getattr(settings, key)
                setattr(settings, key, coerce_value(current, value))
                applied[key] = value

        return ok_response({
            "object_name": obj.name,
            "modifier": "Softbody",
            "applied": applied,
        })

    except Exception as e:
        return error_response(f"Failed to add Soft Body modifier to '{object_name}': {e}")


def _handle_collision_add(params):
    """
    Add a Collision modifier to a mesh object.
    Route: POST /api/physics/collision-add
    """
    object_name = params.get("object_name")
    thickness = params.get("thickness")
    friction = params.get("friction")

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    obj, err = _get_mesh_object(object_name)
    if err:
        return err

    if obj.modifiers.get("Collision") is not None:
        return error_response(
            f"Object '{object_name}' already has a Collision modifier."
        )

    try:
        ensure_context_for_object(obj)

        with temp_override("VIEW_3D"):
            bpy.ops.object.modifier_add(type="COLLISION")

        col_mod = obj.modifiers.get("Collision")
        if col_mod is None:
            return error_response("Failed to add Collision modifier.")

        col_settings = col_mod.settings

        if thickness is not None:
            col_settings.thickness_outer = float(thickness)

        if friction is not None:
            # In Blender, cloth friction on the collision object is 'friction_factor'
            # or 'cloth_friction' depending on version — try both
            if hasattr(col_settings, "cloth_friction"):
                col_settings.cloth_friction = float(friction)
            elif hasattr(col_settings, "friction_factor"):
                col_settings.friction_factor = float(friction)

        return ok_response({
            "object_name": obj.name,
            "modifier": "Collision",
            "thickness_outer": col_settings.thickness_outer,
        })

    except Exception as e:
        return error_response(f"Failed to add Collision modifier to '{object_name}': {e}")


def _handle_fluid_add(params):
    """
    Add a Fluid (Mantaflow) modifier to an object and configure its type.
    Route: POST /api/physics/fluid-add
    """
    object_name = params.get("object_name")
    fluid_type = params.get("type", "").upper()

    if not object_name:
        return error_response("Parameter 'object_name' is required.")

    valid_types = ("DOMAIN", "FLOW", "EFFECTOR")
    if fluid_type not in valid_types:
        return error_response(
            f"Parameter 'type' is required. Valid: {', '.join(valid_types)}."
        )

    obj, err = _get_object(object_name)
    if err:
        return err

    if obj.modifiers.get("Fluid") is not None:
        return error_response(
            f"Object '{object_name}' already has a Fluid modifier."
        )

    try:
        ensure_context_for_object(obj)

        with temp_override("VIEW_3D"):
            bpy.ops.object.modifier_add(type="FLUID")

        fluid_mod = obj.modifiers.get("Fluid")
        if fluid_mod is None:
            return error_response("Failed to add Fluid modifier.")

        fluid_mod.fluid_type = fluid_type

        return ok_response({
            "object_name": obj.name,
            "modifier": "Fluid",
            "fluid_type": fluid_mod.fluid_type,
        })

    except Exception as e:
        return error_response(f"Failed to add Fluid modifier to '{object_name}': {e}")


def _handle_bake(params):
    """
    Bake physics simulations to cache.
    Route: POST /api/physics/bake
    """
    object_name = params.get("object_name")
    start_frame = params.get("start_frame")
    end_frame = params.get("end_frame")

    try:
        if object_name:
            obj, err = _get_object(object_name)
            if err:
                return err
            ensure_context_for_object(obj)

        scene = bpy.context.scene

        if start_frame is not None:
            scene.frame_start = int(start_frame)
        if end_frame is not None:
            scene.frame_end = int(end_frame)

        with temp_override("VIEW_3D"):
            bpy.ops.ptcache.bake_all(bake=True)

        return ok_response({
            "object_name": object_name,
            "baked": True,
            "frame_start": scene.frame_start,
            "frame_end": scene.frame_end,
        })

    except Exception as e:
        return error_response(f"Physics bake failed: {e}")


def _handle_free_bake(params):
    """
    Free (clear) physics bake cache data.
    Route: POST /api/physics/free-bake
    """
    object_name = params.get("object_name")

    try:
        if object_name:
            obj, err = _get_object(object_name)
            if err:
                return err
            ensure_context_for_object(obj)

        with temp_override("VIEW_3D"):
            bpy.ops.ptcache.free_bake_all()

        return ok_response({
            "object_name": object_name,
            "freed": True,
        })

    except Exception as e:
        return error_response(f"Physics free-bake failed: {e}")


# ─── Register routes ─────────────────────────────────────────────────────────────

register_handler("physics", "cloth-add", _handle_cloth_add)
register_handler("physics", "cloth-configure", _handle_cloth_configure)
register_handler("physics", "cloth-pin", _handle_cloth_pin)
register_handler("physics", "rigid-body-add", _handle_rigid_body_add)
register_handler("physics", "rigid-body-configure", _handle_rigid_body_configure)
register_handler("physics", "soft-body-add", _handle_soft_body_add)
register_handler("physics", "collision-add", _handle_collision_add)
register_handler("physics", "fluid-add", _handle_fluid_add)
register_handler("physics", "bake", _handle_bake)
register_handler("physics", "free-bake", _handle_free_bake)
